### PR TITLE
Move translation mark validation from wmllint to schema

### PR DESCRIPTION
**Pull request by CelticMinstrel**
With this, the schema validator can now distinguish between a translatable and a non-translatable string.

* The `t_string` type, which must be translatable, is now a schema built-in type (the only built-in).
* Type definitions now support an `allow_translatable` key, which defaults to false, meaning the value _must not_ be translatable. If set to true, the value is allowed to be translatable. Translatable values in this key will still be matched against the pattern.

### CODE REVIEW

File: data/gui/window/network_transmission.cfg [hunk 1]
```diff
@@ -95,7 +95,6 @@
 									definition = "gold_small"
 									# NOTE: needed to avoid explicit calls to invalidate_layout()
```
Comment (stevecotton) [inline data/gui/window/network_transmission.cfg:96]: This leaves a dangling comment that makes no sense without the line that it referred to.

File: data/core/terrain.cfg [hunk 2]
```diff
@@ -1276,7 +1276,9 @@ Occasionally caves are <italic>text='illuminated'</italic>."
     #editor_image=cave/earthy-wall-hewn-tile
     id=cavewall_earthy_hewn
     name= _ "Cave Wall"
-    editor_name="Reinforced Earthy Cave Wall" # wmllint: ignore
+#ifndef SCHEMA_SHOULD_SKIP_THIS
+    editor_name="Reinforced Earthy Cave Wall"
```
Comment (stevecotton) [inline data/core/terrain.cfg:1280]: This should be translatable.

File: data/core/units/dunefolk/Wyvern_Rider.cfg [hunk 3]
```diff
@@ -14,7 +14,7 @@
     {AMLA_DEFAULT}
     cost=102
     usage=scout
-    description=""
+    #description=""
```
Comment (stevecotton) [inline data/core/units/dunefolk/Wyvern_Rider.cfg:17]: I feel a `[unit_type]` that lacks a description should fail validation.

File: utils/CI/schema_validation.sh [hunk 4]
```diff
@@ -8,7 +8,7 @@ function validate_core()
     echo "------"
     echo "Validating $NAME..."
     
-    ./wesnoth --validate data/_main.cfg &> temp.log || SUCCESS="No"
+    ./wesnoth --validate data/_main.cfg --preprocess-defines=SCHEMA_SHOULD_SKIP_THIS &> temp.log || SUCCESS="No"
```
Comment (soliton-) [inline utils/CI/schema_validation.sh:11]: I'd rather call the define something like `SCHEMA_VALIDATION`.

File: data/core/terrain.cfg (hunk 2 above)
Comment (CelticMinstrel) [reply data/core/terrain.cfg:1280]: The reason I left it untranslated was because it's deprecated… although if there's a non-deprecated terrain with the same name, I may as well just make it translated…

File: data/core/units/dunefolk/Wyvern_Rider.cfg (hunk 3 above)
Comment (CelticMinstrel) [reply data/core/units/dunefolk/Wyvern_Rider.cfg:17]: My thought was that maybe I should just make it consider an empty string without translation mark to be a valid t_string. Does that make sense?

**Comment by CelticMinstrel** [comment]
I'm thinking I'll squash this on merging, so I split the commits up into easier-to-review chunks in case anyone wants to limit their review to a particular subset.

File: data/schema/core/conditionalwml.cfg [hunk 5]
```diff
@@ -25,17 +25,17 @@
...
+		{SIMPLE_KEY equals any}
+		{SIMPLE_KEY not_equals any}
+		{SIMPLE_KEY numerical_equals s_int}
+		{SIMPLE_KEY numerical_not_equals s_int}
+		{SIMPLE_KEY greater_than s_int}
+		{SIMPLE_KEY greater_than_equal_to s_int}
+		{SIMPLE_KEY less_than s_int}
+		{SIMPLE_KEY less_than_equal_to s_int}
+		{SIMPLE_KEY boolean_equals s_bool}
+		{SIMPLE_KEY boolean_not_equals s_bool}
```
Comment (CelticMinstrel) [inline data/schema/core/conditionalwml.cfg:38]: Changing these to s_int and s_bool should maybe be a separate commit? It has nothing to do with the tstring stuff.

File: data/schema/types/basic.cfg [hunk 6]
```diff
@@ -25,14 +25,14 @@
...
-	name=string
+	name=any
 	value=".*"
+	allow_translatable=yes
 [/type]
 [type]
-	name=t_string
-	value="_?.*"
+	name=string
+	value=".*"
```
Comment (CelticMinstrel) [inline data/schema/types/basic.cfg:34]: Note to self: I should swap the order of these types for a better diff.

File: data/campaigns/tutorial/scenarios/01_Tutorial_part_1.cfg [hunk 7]
```diff
@@ -138,8 +138,63 @@
...
+        #  [/set_menu_item]
+        #  [lua]
+        #      code=<<
+        #          wesnoth.interface.set_menu_item(
+        #              'test',
+        #              {description="Test"},
+        #              function() gui.alert("Even this works!") end
+        #          )
+        #      >>
+        #  [/lua]
```
Comment (CelticMinstrel) [inline data/campaigns/tutorial/scenarios/01_Tutorial_part_1.cfg:197]: Whoops, this is old test code from the events PR, it shouldn't be in here.

File: data/campaigns/Descent_Into_Darkness/utils/abilities.cfg [hunk 8]
```diff
@@ -714,9 +713,8 @@
 
                 [floating_text]
                     x,y=$cunit.x,$cunit.y
-                    # wmllint: markcheck off - nothing to translate here
-                    text="<span color='#ff0000'>" + "2" + "</span>"
-                    # wmllint: markcheck on
+                    color=255,0,0
+                    text=2
```
Comment (stevecotton) [inline data/campaigns/Descent_Into_Darkness/utils/abilities.cfg:717]: While all the places changed in this PR use a single color, this wouldn't be able to handle more complex formatting. For example, a hitpoints-to-gold or hitpoints-to-magic_points conversion that has a red negative and a yellow or blue positive number.

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg [hunk 9]
```diff
@@ -6,7 +6,7 @@
     map_generation=lua
     [generator]
         id="cavegen"
-        config_name="Lua Cave Generator"
+        config_name=_"Sceptre of Fire Caves"
```
Comment (stevecotton) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg:9]: Where is this shown?

File: data/test/macros/wml_unit_test_macros.cfg [hunk 10]
```diff
@@ -71,7 +71,7 @@ human#endarg
 Elvish Archer#endarg
 
     [test]
-        name = "Unit Test " + {NAME}
+        name = _ "Unit Test " + {NAME}
```
Comment (stevecotton) [inline data/test/macros/wml_unit_test_macros.cfg:74]: This doesn't feel like a good change - if the schema is checking for a mark, but isn't complaining about the untranslatable `NAME`.

**Comment by stevecotton** [comment]
I think the changes to Dead Water and SotBE are bugfixes that should be cherry-picked.

File: data/campaigns/Descent_Into_Darkness/utils/abilities.cfg (hunk 8 above)
Comment (CelticMinstrel) [reply data/campaigns/Descent_Into_Darkness/utils/abilities.cfg:717]: I don't really get your example, but you're correct that more complex formatting could just include the translation mark.

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg (hunk 9 above)
Comment (CelticMinstrel) [reply data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg:9]: I don't think it's shown anywhere, because this file is not parsed in map editor mode.

If it _was_ parsed in map editor mode, you would see it in the Generate Map dialog.

File: data/test/macros/wml_unit_test_macros.cfg (hunk 10 above)
Comment (CelticMinstrel) [reply data/test/macros/wml_unit_test_macros.cfg:74]: A value is considered translatable if any segment of it is translatable, and it's fairly common to concatenate variously-translatable parts in this way (for example, untranslatable formatting codes and translatable content). This at least avoids the trap of putting a macro inside the translatable string, and since this is in wesnoth-test, it won't be sent to the translators anyway.

File: data/gui/window/debug_clock.cfg [hunk 11]
```diff
@@ -149,7 +149,7 @@
 						minimum_value = 0
 						maximum_value = 3599
 						step_size = 1
-						label = 3599
+						label = _ "3599"
```
Comment (stevecotton) [inline data/gui/window/debug_clock.cfg:152]: Should `[slider]label=` be translatable? It looks like it needs to be a number.

File: data/test/maps/pathfind_1.cfg [hunk 12]
```diff
@@ -61,7 +61,6 @@ victory_when_enemies_defeated=yes
     lawful_bonus=0
     mask=""
     #textdomain wesnoth-help
```
Comment (stevecotton) [inline data/test/maps/pathfind_1.cfg:63]: Is this line still necessary?

File: data/gui/window/debug_clock.cfg (hunk 11 above)
Comment (CelticMinstrel) [reply data/gui/window/debug_clock.cfg:152]: I think it doesn't _need_ to be a number but usually will be… it also won't usually be set in the WML, since it needs to update dynamically as the slider moves.

This is the debug clock, so I don't really care if it's a bit weird, but I can change something if that seems like a good idea.

File: data/test/maps/pathfind_1.cfg (hunk 12 above)
Comment (CelticMinstrel) [reply data/test/maps/pathfind_1.cfg:63]: Nope!

File: utils/CI/schema_validation.sh (hunk 4 above)
Comment (CelticMinstrel) [reply utils/CI/schema_validation.sh]: Done, and I also defined it in the C++ instead of in the CI script.

File: src/theme.cpp [hunk 13]
```diff
@@ -431,7 +431,7 @@ theme::label::label()
 
 theme::label::label(std::size_t sw, std::size_t sh, const config& cfg)
 	: object(sw, sh, cfg)
-	, text_(cfg["prefix"].str() + cfg["text"].str() + cfg["postfix"].str())
+	, text_(cfg["prefix"].str() + cfg["prefix_literal"].str() + cfg["text"].str() + cfg["postfix_literal"].str() + cfg["postfix"].str())
```
Comment (Pentarctagon) [inline src/theme.cpp:434]: What is prefix_literal and postfix_literal?

File: src/gui/dialogs/network_transmission.cpp [hunk 14]
```diff
@@ -113,6 +113,9 @@ void network_transmission::pre_show(window& window)
 		subtitle_label.set_use_markup(true);
 	}
 
+	// NOTE: needed to avoid explicit calls to invalidate_layout()
+	// in network_transmission::pump_monitor::process()
+	find_widget<label>(&window, "numeric_progress", false).set_label(" ");
```
Comment (Pentarctagon) [inline src/gui/dialogs/network_transmission.cpp:118]: What is this doing?

File: src/theme.cpp (hunk 13 above)
Comment (CelticMinstrel) [reply src/theme.cpp:434]: It's pretty much exactly what it says – a literal, untranslated string that prefixes or postfixes the main string.

File: src/gui/dialogs/network_transmission.cpp (hunk 14 above)
Comment (CelticMinstrel) [reply src/gui/dialogs/network_transmission.cpp:118]: I honestly have no idea. The WML set the label to a single space with this comment, but that makes it fail validation, so I figured setting it in the C++ would have the same effect.

File: src/theme.cpp (hunk 13 above)
Comment (Pentarctagon) [reply src/theme.cpp:434]: Why is that being added with this PR?

File: src/gui/dialogs/network_transmission.cpp (hunk 14 above)
Comment (Pentarctagon) [reply src/gui/dialogs/network_transmission.cpp:118]: Given there's no longer any explicit call to invalidate_layout in network_transmission::pump_monitor::process(), is it even still needed?

File: src/theme.cpp (hunk 13 above)
Comment (CelticMinstrel) [reply src/theme.cpp:434]: It was so that certain things in the theme didn't have to be translated. In particular I think there were several cases of strings consisting of only spaces.

File: src/gui/dialogs/network_transmission.cpp (hunk 14 above)
Comment (CelticMinstrel) [reply src/gui/dialogs/network_transmission.cpp:118]: That's a good question! The comment suggests removing this would mean we need to add explicit calls there, but I have no idea why a single space would make that difference.

Comment (Pentarctagon) [reply src/gui/dialogs/network_transmission.cpp:118]: Looks like https://github.com/wesnoth/wesnoth/commit/34d45d3ee5ff4d3ec939e5ae0272a292eefce60f added this in WML.

**Review by Pentarctagon — APPROVED** [review]
Overall I guess this seems like pretty uncontroversial change.  If this is ever removed from wmllint as well, then at that point I'd expect there to be more discussion about schema being accessible, etc.
